src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.CreateToken.cs

```diff
@@ -0,0 +1,1383 @@
+                    encodedChars[sizeOfEncodedHeader] = '.';
+                    int sizeOfEncodedPayload = Base64UrlEncoder.Encode(utf8ByteMemoryStream.GetBuffer().AsSpan(headerLength, payloadEnd - headerLength), encodedChars.AsSpan(sizeOfEncodedHeader + 1));
+                    // encodeChars => 'EncodedHeader.EncodedPayload'
+
+                    // Get ASCII Bytes of 'EncodedHeader.EncodedPayload' which is used to calculate the signature
+                    asciiBytes = ArrayPool<byte>.Shared.Rent(Encoding.ASCII.GetMaxByteCount(encodedBufferSize));
+                    int sizeOfEncodedHeaderAndPayloadAsciiBytes
+                        = Encoding.ASCII.GetBytes(encodedChars, 0, sizeOfEncodedHeader + sizeOfEncodedPayload + 1, asciiBytes, 0);
+
+                    encodedChars[sizeOfEncodedHeader + sizeOfEncodedPayload + 1] = '.';
+                    // encodedChars => 'EncodedHeader.EncodedPayload.'
+
+                    int sizeOfEncodedSignature = 0;
+                    if (tokenDescriptor.SigningCredentials != null)
+                    {
+#if NET6_0_OR_GREATER
+                        signatureBytes = ArrayPool<byte>.Shared.Rent(signatureSize);
+                        bool signatureSucceeded = JwtTokenUtilities.CreateSignature(
+                            asciiBytes.AsSpan(0, sizeOfEncodedHeaderAndPayloadAsciiBytes),
+                            signatureBytes,
+                            tokenDescriptor.SigningCredentials,
+                            out int signatureLength);
+#else
+                        signatureBytes = JwtTokenUtilities.CreateEncodedSignature(asciiBytes, 0, sizeOfEncodedHeaderAndPayloadAsciiBytes, tokenDescriptor.SigningCredentials);
+                        int signatureLength = signatureBytes.Length;
+#endif
+                        sizeOfEncodedSignature = Base64UrlEncoder.Encode(signatureBytes.AsSpan(0, signatureLength), encodedChars.AsSpan(sizeOfEncodedHeader + sizeOfEncodedPayload + 2));
+                    }
+
+                    if (tokenDescriptor.EncryptingCredentials != null)
+                    {
+                        return EncryptToken(
+                            Encoding.UTF8.GetBytes(encodedChars, 0, sizeOfEncodedHeader + sizeOfEncodedPayload + sizeOfEncodedSignature + 2),
+                            tokenDescriptor.EncryptingCredentials,
+                            tokenDescriptor.CompressionAlgorithm,
+                            tokenDescriptor.AdditionalHeaderClaims,
+                            tokenDescriptor.TokenType);
+                    }
+                    else
+                    {
+                        return encodedChars.AsSpan(0, sizeOfEncodedHeader + sizeOfEncodedPayload + sizeOfEncodedSignature + 2).ToString();
+                    }
+                }
+                finally
+                {
+                    if (encodedChars is not null)
```

Comment:
for these three, did you look at the average cases sizes and check if any of them make sense to stackalloc?

Reply:
@keegan-caruso another great idea.
Let's put it in our TODO list.